components/timezone/src/provider/names.rs

```diff
@@ -0,0 +1,207 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! 🚧 \[Unstable\] Property names-related data for this component
+//!
+//! <div class="stab unstable">
+//! 🚧 This code is considered unstable; it may change at any time, in breaking or non-breaking ways,
+//! including in SemVer minor releases. While the serde representation of data structs is guaranteed
+//! to be stable, their Rust representation might not be. Use with caution.
+//! </div>
+//!
+//! Read more about data providers: [`icu_provider`]
+
+use alloc::boxed::Box;
+use core::cmp::Ordering;
+use core::str;
+
+use icu_provider::prelude::*;
+
+use crate::TimeZoneBcp47Id;
+use tinystr::UnvalidatedTinyAsciiStr;
+use zerovec::ule::{UnvalidatedStr, VarULE};
+use zerovec::{maps::ZeroMapKV, VarZeroSlice, VarZeroVec, ZeroMap};
+
+/// This is a time zone identifier that can be "loose matched" as according to
+/// [ECMAScript Temporal](https://tc39.es/proposal-temporal/#sec-isavailabletimezonename)
+///
+/// (matched case-insensitively in ASCII)
+///
+/// This is expected to be ASCII, but we do not rely on this invariant anywhere except during
+/// datagen.
+///
+/// The Ord impl will sort things using strict equality, but in such a way that all loose-equal items
+/// will sort into the same area, such that a map can be searched for both strict and loose equality.
+///
+/// <div class="stab unstable">
+/// 🚧 This code is considered unstable; it may change at any time, in breaking or non-breaking ways,
+/// including in SemVer minor releases. While the serde representation of data structs is guaranteed
+/// to be stable, their Rust representation might not be. Use with caution.
+/// </div>
+///
+/// # Examples
+///
+/// Using a [`NormalizedTimeZoneIdStr`] as the key of a [`ZeroMap`]:
+///
+/// ```
+/// use icu_timezone::provider::names::NormalizedTimeZoneIdStr;
+/// use zerovec::ZeroMap;
+///
+/// let map: ZeroMap<NormalizedTimeZoneIdStr, usize> = [
+///     (NormalizedTimeZoneIdStr::from_str("America/Los_Angeles"), 11),
+///     (NormalizedTimeZoneIdStr::from_str("Asia/Kolkata"), 22),
+///     (NormalizedTimeZoneIdStr::from_str("Europe/Berlin"), 33),
+/// ]
+/// .into_iter()
+/// .collect();
+///
+/// let key_approx = NormalizedTimeZoneIdStr::from_str("europe/berlin");
+/// let key_exact = NormalizedTimeZoneIdStr::from_str("Europe/Berlin");
+///
+/// // Strict lookup:
+/// assert_eq!(None, map.get_copied(key_approx));
+/// assert_eq!(Some(33), map.get_copied(key_exact));
+///
+/// // Loose lookup:
+/// assert_eq!(Some(33), map.get_copied_by(|u| u.cmp_loose(key_approx)));
+/// assert_eq!(Some(33), map.get_copied_by(|u| u.cmp_loose(key_exact)));
+/// ```
+#[derive(PartialEq, Eq)] // VarULE wants these to be byte equality
+#[derive(Debug, VarULE)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize))]
+#[repr(transparent)]
```

Comment:
thought: we really should see if we can get someone to make the `make_wrapper_varule` macro